exp/topology/desiredstate/desired_state_test.go

```diff
@@ -3545,25 +3572,46 @@ func TestCalculateRefDesiredAPIVersion(t *testing.T) {
 	}
 }
 
-func validateCleanupCluster(req runtimehooksv1.RequestObject) error {
-	var cluster clusterv1beta1.Cluster
-	switch req := req.(type) {
-	case *runtimehooksv1.BeforeClusterUpgradeRequest:
-		cluster = req.Cluster
-	case *runtimehooksv1.AfterControlPlaneUpgradeRequest:
-		cluster = req.Cluster
-	default:
-		return fmt.Errorf("unhandled request type %T", req)
-	}
+func validateCleanupCluster(originalCluster *clusterv1.Cluster) func(req runtimehooksv1.RequestObject) error {
+	return func(req runtimehooksv1.RequestObject) error {
+		var cluster clusterv1beta1.Cluster
+		switch req := req.(type) {
+		case *runtimehooksv1.BeforeClusterUpgradeRequest:
+			cluster = req.Cluster
+		case *runtimehooksv1.AfterControlPlaneUpgradeRequest:
+			cluster = req.Cluster
+		default:
+			return fmt.Errorf("unhandled request type %T", req)
+		}
 
-	if cluster.GetManagedFields() != nil {
-		return errors.New("managedFields should have been cleaned up")
-	}
-	if _, ok := cluster.Annotations[corev1.LastAppliedConfigAnnotation]; ok {
-		return errors.New("last-applied-configuration annotation should have been cleaned up")
-	}
-	if _, ok := cluster.Annotations[conversion.DataAnnotation]; ok {
-		return errors.New("conversion annotation should have been cleaned up")
+		if cluster.GetManagedFields() != nil {
+			return errors.New("managedFields should have been cleaned up")
+		}
+		if _, ok := cluster.Annotations[corev1.LastAppliedConfigAnnotation]; ok {
+			return errors.New("last-applied-configuration annotation should have been cleaned up")
+		}
+		if _, ok := cluster.Annotations[conversion.DataAnnotation]; ok {
+			return errors.New("conversion annotation should have been cleaned up")
+		}
+
+		v1beta2Cluster := &clusterv1.Cluster{}
+		if err := cluster.ConvertTo(v1beta2Cluster); err != nil {
+			return err
+		}
+
+		// Apply same changes that are expected for a cluster sent to runtime hook.
+		originalClusterCopy := originalCluster.DeepCopy()
+		originalClusterCopy.SetManagedFields(nil)
+		if originalClusterCopy.Annotations != nil {
+			annotations := maps.Clone(cluster.Annotations)
+			delete(annotations, corev1.LastAppliedConfigAnnotation)
+			delete(annotations, conversion.DataAnnotation)
+			originalClusterCopy.Annotations = annotations
+		}
+
+		if !apiequality.Semantic.DeepEqual(originalClusterCopy, v1beta2Cluster) {
+			return errors.Errorf("call to extension is not passing the expected cluster object: %s", cmp.Diff(originalClusterCopy, v1beta2Cluster))
+		}
+		return nil
```

Comment:
Is this the right place to put this? The function name is throwing me off a bit?